### PR TITLE
[LAP-66] Feature/glTF loading

### PR DESCRIPTION
**Pull request by Jelled1st**
adds glb file loading (binary gltf)

### CODE REVIEW

File: legion/engine/core/data/importers/mesh_importers.cpp [hunk 1]
```diff
@@ -36,6 +40,113 @@ namespace legion::core::detail
...
+                    data->at(dataIndex++) = math::color(r, g, b, a);
+                }
+            }
+            else
+            {
+                log::warn("Warning: Vert colors were not stored as UNSIGNED BYTE/SHORT or float, skipping");
+            }
+        }
+        else log::warn("Warning: Vert colors were not vec3 or vec4, skipping colors");
+    }
```
Comment (Developer-The-Great) [inline legion/engine/core/data/importers/mesh_importers.cpp]: I think some refactoring  and or commenting would be helpfull

File: legion/engine/core/data/importers/mesh_importers.cpp [hunk 2]
```diff
@@ -180,4 +291,221 @@ namespace legion::core
...
+            uint i1 = meshData.indices[i + 1];
+            uint i2 = meshData.indices[i + 2];
+            meshData.indices[i + 1] = i2;
+            meshData.indices[i + 2] = i1;
+        }
+
+        mesh::calculate_tangents(&meshData);
+
+        return decay(Ok(meshData));
+    }
```
Comment (Developer-The-Great) [inline legion/engine/core/data/importers/mesh_importers.cpp]: also might need a refactor

**Review by Developer-The-Great — CHANGES_REQUESTED** [review]
needs refactoring in some places

**Review by Developer-The-Great — APPROVED** [review]

File: deps/include/tinygltf/stb_image.h [hunk 3]
```diff
@@ -0,0 +1,7530 @@
+/* stb_image - v2.21 - public domain image loader - http://nothings.org/stb
+                                  no warranty implied; use at your own risk
+
+   Do this:
+      #define STB_IMAGE_IMPLEMENTATION
+   before you include this file in *one* C or C++ file to create the implementation.
+
+   // i.e. it should look like this:
+   #include ...
+   #include ...
```
Comment (GlynLeine) [inline deps/include/tinygltf/stb_image.h]: are we including this header twice now?

File: legion/engine/core/data/importers/mesh_importers.cpp [hunk 4]
```diff
@@ -36,6 +40,141 @@ namespace legion::core::detail
...
+     * \param bufferView - the tinygltf::BufferView containing information about data size and offset
+     * \param data - std::Vector<T> where the buffer is going to be copied into. The vector will be resized to vector.size()+(tinygltf data size)
+     */
+    template <class T>
+    void handleGltfBuffer(const tinygltf::Buffer& buffer, const tinygltf::BufferView& bufferView, std::vector<T>* data)
+    {
+        /*for (size_t i = 0; i < bufferView.byteLength; i += sizeof(T))
+        {
+            data->push_back(*reinterpret_cast<const T*>(&buffer.data.at(i) + bufferView.byteOffset));
+        }*/
```
Comment (GlynLeine) [inline legion/engine/core/data/importers/mesh_importers.cpp]: cleanup commented out code?

File: legion/engine/core/data/importers/mesh_importers.cpp [hunk 5]
```diff
@@ -36,6 +40,141 @@ namespace legion::core::detail
...
+            {
+                // Vertex colors in unsigned short
+                // Currently not supported
+                log::debug("Color component type: u short");
+                log::warn("Warning: Vert colors for UNSIGNED SHORT not implemented");
+            }
+            else if (componentType == TINYGLTF_COMPONENT_TYPE_FLOAT)
+            {
+                // Vertex colors in float
+                log::debug("Color component type: float");
```
Comment (GlynLeine) [inline legion/engine/core/data/importers/mesh_importers.cpp]: might wanna remove the debug logs or turn them into trace logs, keep the warnings tho

File: legion/engine/core/data/importers/mesh_importers.cpp [hunk 6]
```diff
@@ -36,6 +40,141 @@ namespace legion::core::detail
...
+            else
+            {
+                log::warn("Warning: Vert colors were not stored as UNSIGNED BYTE/SHORT or float, skipping");
+            }
+        }
+        else if (accessorType == TINYGLTF_TYPE_VEC4)
+        {
+            // Copy the vertex colors into the vector, keeping in my mind that the vertex color data is r,g,b,a
+
+            log::debug("Color accessor type: Vec4");
```
Comment (GlynLeine) [inline legion/engine/core/data/importers/mesh_importers.cpp]: same story, either remove debug log or turn them into trace logs, this is not interesting to a gameplay programmer.

File: legion/engine/core/data/importers/mesh_importers.cpp [hunk 7]
```diff
@@ -180,4 +319,244 @@ namespace legion::core
...
+
+        size_t offset = 0;
+        core::mesh meshData;
+        for (auto& mesh : model.meshes)
+        {
+            // Loop through all the meshes (submeshes)
+
+            sub_mesh m;
+            m.name = mesh.name;
+            log::debug("Loading submesh: {}", m.name);
```
Comment (GlynLeine) [inline legion/engine/core/data/importers/mesh_importers.cpp]: same story

File: legion/engine/core/data/importers/mesh_importers.cpp [hunk 8]
```diff
@@ -180,4 +319,244 @@ namespace legion::core
...
+                        detail::handleGltfBuffer<math::vec2>(buff, view, &(meshData.uvs));
+                    }
+                    else if (attrib.first.compare("COLOR_0") == 0)
+                    {
+                        // Vertex color data
+                        detail::handleGltfVertexColor(buff, view, accessor.type, accessor.componentType, &(meshData.colors));
+                    }
+                    else
+                    {
+                        log::warn("\n\n\n\nMore data to be found in .gbl. Data can be accesed through: {}\n\n\n", attrib.first);
```
Comment (GlynLeine) [inline legion/engine/core/data/importers/mesh_importers.cpp]: why all the `\n`???

File: legion/engine/core/data/importers/mesh_importers.cpp [hunk 9]
```diff
@@ -180,4 +319,244 @@ namespace legion::core
...
+
+        if (!err.empty())
+        {
+            // Check and print errors
+            log::error("Error: {}", err);
+        }
+        if (!warn.empty())
+        {
+            // Check and print warnings
+            log::warn("Warning: {}", warn);
```
Comment (GlynLeine) [inline legion/engine/core/data/importers/mesh_importers.cpp]: error and warning already do the whole "error:" thing

File: legion/engine/core/data/importers/mesh_importers.cpp [hunk 10]
```diff
@@ -180,4 +319,244 @@ namespace legion::core
...
+
+        size_t offset = 0;
+        core::mesh meshData;
+        for (auto& mesh : model.meshes)
+        {
+            // Loop through all the meshes (submeshes)
+
+            sub_mesh m;
+            m.name = mesh.name;
+            log::debug("Loading submesh: {}", m.name);
```
Comment (GlynLeine) [inline legion/engine/core/data/importers/mesh_importers.cpp]: same story as before

File: legion/engine/core/data/importers/mesh_importers.cpp [hunk 11]
```diff
@@ -180,4 +319,244 @@ namespace legion::core
...
+                    // Depending on the attribute the data is copied into a different std::vector in meshData
+
+                    const tg::Accessor& accessor = model.accessors.at(attrib.second);
+                    tg::BufferView& view = model.bufferViews.at(accessor.bufferView);
+                    tg::Buffer& buff = model.buffers.at(view.buffer);
+                    if (attrib.first.compare("POSITION") == 0)
+                    {
+                        // Position data
+                        detail::handleGltfBuffer<math::vec3>(buff, view, &(meshData.vertices));
+                        log::debug("Vertices count: {}", meshData.vertices.size());
```
Comment (GlynLeine) [inline legion/engine/core/data/importers/mesh_importers.cpp]: same story as before

File: legion/engine/core/data/importers/mesh_importers.cpp [hunk 12]
```diff
@@ -180,4 +319,244 @@ namespace legion::core
...
+                        detail::handleGltfBuffer<math::vec2>(buff, view, &(meshData.uvs));
+                    }
+                    else if (attrib.first.compare("COLOR_0") == 0)
+                    {
+                        // Vertex color data
+                        detail::handleGltfVertexColor(buff, view, accessor.type, accessor.componentType, &(meshData.colors));
+                    }
+                    else
+                    {
+                        log::warn("\n\n\n\nMore data to be found in .gbl. Data can be accesed through: {}\n\n\n", attrib.first);
```
Comment (GlynLeine) [inline legion/engine/core/data/importers/mesh_importers.cpp]: `\n\n\n\n`???

**Review by GlynLeine — CHANGES_REQUESTED** [review]

File: legion/engine/core/data/importers/mesh_importers.cpp [hunk 13]
```diff
@@ -36,6 +40,141 @@ namespace legion::core::detail
             return hash != other.hash;
         }
     };
+
+    /**
+     * @brief Function to copy tinygltf buffer data into the correct mesh data vector
+     * 
+     * \param buffer - The tinygltf::Buffer buffer containing the data
+     * \param bufferView - the tinygltf::BufferView containing information about data size and offset
+     * \param data - std::Vector<T> where the buffer is going to be copied into. The vector will be resized to vector.size()+(tinygltf data size)
```
Comment (GlynLeine) [inline legion/engine/core/data/importers/mesh_importers.cpp]: We use `@param` not `\param`

**Review by GlynLeine — APPROVED** [review]

**Pull request merged**